### ISSUE TITLE
Feature request: Legend for imshow

### ISSUE THREAD

**Issue by cool-RR** [matplotlib/matplotlib#18334]
I'm using imshow to show a 3x3 array of floating point numbers and it's very hard to figure out. Here's the image:

![](https://i.imgur.com/jUCQj8A.png)

Can people who view this plot really figure out whether purple is high or low? And what the numbers are?

It looks like [this question](https://stackoverflow.com/questions/25482876/how-to-add-legend-to-imshow-in-matplotlib) collects a few recipes for creating legends for imshow. I wish there was a `legend=True` argument that just makes imshow create such a legend automatically.

**Comment by jklymak** [comment]
Usually `imshow` is used with a colorbar. https://matplotlib.org/3.1.0/gallery/color/colorbar_basics.html

**Comment by cool-RR** [comment]
That's perfect, thank you.

**Comment by cool-RR** [comment]
@jklymak I closed the PR, but I'd still like to ask you a general question, as someone who isn't very experienced with matplotlib.

I noticed that matplotlib is very imperative, in contrast to declarative. For example, when you want to show an `imshow` plot with a legend like in this example, you need to first create the plot, then add the legend, then show the plot. In more complicated plots, you could be doing 10+ actions one after another to create a plot.

I would find a declarative interface to be much more elegant, i.e. one call to create the plot and lots of keyword arguments, some of them nested, to specify all the options. Of course, matplotlib being a very mature project, the current interface must be maintained forever. But, do you know of any such declarative interface to matplotlib? Either as part of the library, or a separate library that uses matplotlib?

**Comment by jklymak** [comment]
I find the declarative interfaces to be very hard to parse and the attempts at elegance break down with any complexity, so you are asking the wrong person to recommend one.

**Comment by cool-RR** [comment]
Cool, I'll ask on the mailing list.
